### PR TITLE
Adding template Snowflake to GCS

### PR DESCRIPTION
**Pull request by varunika**
Added Template Snowflake to GCS. Tested for all possible inputs i.e. json, parquet, avro, csv.

### CODE REVIEW

File: python/dataproc_templates/snowflake/README.md [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+--snowflake.to.gcs.sf.query=<snowflake-select-query> \
+--snowflake.to.gcs.output.location=<gcs-output-location> \
+--snowflake.to.gcs.output.format=<csv|avro|orc|json|parquet> \
+--snowflake.to.gcs.output.mode=<Overwrite|ErrorIfExists|Append|Ignore> \
+--snowflake.to.gcs.partition.column=<gcs-output-partitionby-columnname> \
+--snowflake.gcs.sf.autopushdown=<on|off>
+```
+
+### Configurable Parameters
+Following properties are available in commandline (`python main.py --template JDBCTOJDBC --help`):
```
Comment (NiloFreitas) [inline python/dataproc_templates/snowflake/README.md:28]: Fix typo: JDBCTOJDBC to SNOWFLAKETOGCS

File: python/dataproc_templates/snowflake/README.md [hunk 2]
```diff
@@ -0,0 +1,133 @@
+## 1. Snowflake To GCS
+
```
Comment (NiloFreitas) [inline python/dataproc_templates/snowflake/README.md:2]: Add Introduction:
Template for reading data from a Snowflake table or custom query and writing Google Cloud Storage. It supports writing JSON, CSV, Parquet and Avro formats.

File: python/test/snowflake/test_snowflake_to_gcs.py [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+             "--snowflake.to.gcs.output.location=gs://test", 
+             "--snowflake.to.gcs.output.format=json", 
+             "--snowflake.to.gcs.sf.autopushdown=yes",
+             "--snowflake.to.gcs.output.mode=append"
+             ]
+        )
+        
+        mock_spark_session.read.format().options().option().load.return_value = mock_spark_session.dataframe.DataFrame
+        snowflake_to_gcs_template.write_data(Logger, mock_parsed_args, mock_spark_session.dataframe.DataFrame)
+        mock_spark_session.dataframe.DataFrame.write.mode.assert_called_with(constants.OUTPUT_MODE_APPEND)
```
Comment (NiloFreitas) [inline python/test/snowflake/test_snowflake_to_gcs.py:300]: Add an "assert_called_with" the json format

File: python/test/snowflake/test_snowflake_to_gcs.py [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+             "--snowflake.to.gcs.output.location=gs://test", 
+             "--snowflake.to.gcs.output.format=parquet", 
+             "--snowflake.to.gcs.sf.autopushdown=yes",
+             "--snowflake.to.gcs.output.mode=append"
+             ]
+        )
+        
+        mock_spark_session.read.format().options().option().load.return_value = mock_spark_session.dataframe.DataFrame
+        snowflake_to_gcs_template.write_data(Logger, mock_parsed_args, mock_spark_session.dataframe.DataFrame)
+        mock_spark_session.dataframe.DataFrame.write.mode.assert_called_with(constants.OUTPUT_MODE_APPEND)
```
Comment (NiloFreitas) [inline python/test/snowflake/test_snowflake_to_gcs.py:274]: Add an "assert_called_with" the parquet format

File: python/test/snowflake/test_snowflake_to_gcs.py [hunk 5]
```diff
@@ -0,0 +1,328 @@
...
+from dataproc_templates.snowflake.snowflake_to_gcs import SnowflakeToGCSTemplate
+import dataproc_templates.util.template_constants as constants
+
+
+class TestSnowflakeToGCSTemplate:
+    """
+    Test suite for SnowflakeToGCSTemplate
+    """
+
+    def test_parse_args_1(self):
```
Comment (NiloFreitas) [inline python/test/snowflake/test_snowflake_to_gcs.py:14]: rename to test_parse_args_query

File: python/test/snowflake/test_snowflake_to_gcs.py [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+        assert parsed_args["snowflake.to.gcs.sf.url"] == "url"
+        assert parsed_args["snowflake.to.gcs.sf.user"] == "user"
+        assert parsed_args["snowflake.to.gcs.sf.password"] == "password"
+        assert parsed_args["snowflake.to.gcs.sf.database"] == "database"
+        assert parsed_args["snowflake.to.gcs.sf.query"] == "select 1"
+        assert parsed_args["snowflake.to.gcs.output.location"] == "gs://test"
+        assert parsed_args["snowflake.to.gcs.output.format"] == "csv"
+        assert parsed_args["snowflake.to.gcs.partition.column"] == "col" 
+        
+    def test_parse_args_2(self):
```
Comment (NiloFreitas) [inline python/test/snowflake/test_snowflake_to_gcs.py:38]: rename to test_parse_args_table

File: python/test/snowflake/test_snowflake_to_gcs.py [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+        assert parsed_args["snowflake.to.gcs.sf.user"] == "user"
+        assert parsed_args["snowflake.to.gcs.sf.password"] == "password"
+        assert parsed_args["snowflake.to.gcs.sf.database"] == "database"
+        assert parsed_args["snowflake.to.gcs.sf.schema"] == "schema"
+        assert parsed_args["snowflake.to.gcs.sf.table"] == "table"
+        assert parsed_args["snowflake.to.gcs.output.location"] == "gs://test"
+        assert parsed_args["snowflake.to.gcs.output.format"] == "csv"
+        assert parsed_args["snowflake.to.gcs.partition.column"] == "col"
+        
+    def test_parse_args_3(self):
```
Comment (NiloFreitas) [inline python/test/snowflake/test_snowflake_to_gcs.py:64]: rename to test_parse_args_warehouse

File: python/dataproc_templates/snowflake/README.md [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+--snowflake.to.gcs.sf.schema=<snowflake-schema> \
+--snowflake.to.gcs.sf.warehouse=<snowflake-warehouse> \
+--snowflake.to.gcs.sf.query=<snowflake-select-query> \
+--snowflake.to.gcs.output.location=<gcs-output-location> \
+--snowflake.to.gcs.output.format=<csv|avro|orc|json|parquet> \
+--snowflake.to.gcs.output.mode=<Overwrite|ErrorIfExists|Append|Ignore> \
+--snowflake.to.gcs.partition.column=<gcs-output-partitionby-columnname> \
+--snowflake.gcs.sf.autopushdown=<on|off>
+```
+
```
Comment (NiloFreitas) [inline python/dataproc_templates/snowflake/README.md:26]: Add Required JAR files section, and explain how to get the snowflake-jdbc-3.13.14.jar and the spark-snowflake_2.12-2.10.0-spark_3.1.jar

File: python/dataproc_templates/snowflake/README.md [hunk 9]
```diff
@@ -0,0 +1,133 @@
+## 1. Snowflake To GCS
+
+General Execution:
+
+```
+export GCP_PROJECT=<gcp-project-id>
+export REGION=<gcp-project-region>
+export GCS_STAGING_LOCATION=<gcs-bucket-staging-folder-path>
+export SUBNET=<gcp-project-dataproc-clusters-subnet>
```
Comment (NiloFreitas) [inline python/dataproc_templates/snowflake/README.md:9]: Add the export JARS="gs://<...>/spark-snowflake_2.12-2.10.0-spark_3.1.jar,gs://<...>/snowflake-jdbc-3.13.14.jar"

File: python/dataproc_templates/snowflake/README.md [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+        ```
+        NOTE: Incase of snowflake.to.gcs.sf.query, if the query contains joins on multiple tables from different schemas, ensure all the schemas are mentioned within the query. As for the snowflake.to.gcs.sf.schema property in this case, you can either not use it at all or provide name of one of the schemas being used in the query.
+
+
+### Example submission
+```
+export GCP_PROJECT="sample-project"
+export REGION="us-central1"
+export SUBNET="default"
+export GCS_STAGING_LOCATION="gs://test-bucket"
```
Comment (NiloFreitas) [inline python/dataproc_templates/snowflake/README.md:119]: Add the export JARS="gs://<...>/spark-snowflake_2.12-2.10.0-spark_3.1.jar,gs://<...>/snowflake-jdbc-3.13.14.jar"

**Review by NiloFreitas — CHANGES_REQUESTED** [review]
Great work Varunika!

File: python/test/snowflake/test_snowflake_to_gcs.py (hunk 4 above)
Comment (varunika) [reply python/test/snowflake/test_snowflake_to_gcs.py:274]: line no. 275 id doing an asser_called_with for parquet. Is it different from what you're expecting?

File: python/test/snowflake/test_snowflake_to_gcs.py (hunk 3 above)
Comment (varunika) [reply python/test/snowflake/test_snowflake_to_gcs.py:300]: Line no.301 is performing an assert_called_with on json(). Is it different from what you're expecting?

File: python/test/snowflake/test_snowflake_to_gcs.py (hunk 4 above)
Comment (NiloFreitas) [reply python/test/snowflake/test_snowflake_to_gcs.py:274]: Sure, you are right, I did not see that. Resolved.

File: python/test/snowflake/test_snowflake_to_gcs.py (hunk 3 above)
Comment (NiloFreitas) [reply python/test/snowflake/test_snowflake_to_gcs.py:300]: Sure, you are right, I did not see that. Resolved.

**Review by NiloFreitas — APPROVED** [review]
Awesome work Varunika! LGTM
Obs: In line 129 of the readme it is written gs:test_bucket instead of gs://test_bucket

**Comment by varunika** [comment]
> Awesome work Varunika! LGTM Obs: In line 129 of the readme it is written gs:test_bucket instead of gs://test_bucket

Noted. Will send in an edit in the next PR.

**Pull request merged**